subsys/ipc/ipc_service_backend/ipc_rpmsg_multi_instance.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (c) 2020-2021, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <ipc/ipc_service_backend.h>
+#include <ipc/rpmsg_multi_instance.h>
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <device.h>
+
+#define LOG_MODULE_NAME ipc_rpmsg_multi_instance
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_IPC_SERVICE_LOG_LEVEL);
+
+#define MASTER IS_ENABLED(CONFIG_RPMSG_MULTI_INSTANCE_MASTER)
+
+#define NUM_ENDPOINTS          CONFIG_RPMSG_MULTI_INSTANCES_NO
+#define WORK_QUEUE_STACK_SIZE  CONFIG_IPC_SERVICE_BACKED_RPMSG_MI_WORK_QUEUE_STACK_SIZE
+#define IPM_CH_NAME_LEN        8
+
+K_THREAD_STACK_ARRAY_DEFINE(ipm_stack_area, NUM_ENDPOINTS, WORK_QUEUE_STACK_SIZE);
+
+/* Internal definition of the structure ipc_ept*/
+struct ipc_ept {
+	struct rpmsg_mi_ept rpmsg_ep;
+	struct ipc_service_cb cb;
+	void *priv;
+};
+
+struct ipc_rpmsg_mi_endpoints {
+	const char          *name;
+	struct rpmsg_mi_ctx ctx;
+	struct ipc_ept      ept;
+	char                ipm_tx_name[IPM_CH_NAME_LEN];
+	char                ipm_rx_name[IPM_CH_NAME_LEN];
+};
+
+static struct ipc_rpmsg_mi_endpoints endpoints[NUM_ENDPOINTS];
+
+static int common_bound_cb(void *priv)
+{
+	struct ipc_ept *ept = (struct ipc_ept *)priv;
+
+	if (ept->cb.bound) {
+		return ept->cb.bound(ept->priv);
+	}
+	return 0;
+}
+
+static int common_recv_cb(void *data, size_t len, void *priv)
+{
+	struct ipc_ept *ept = (struct ipc_ept *)priv;
+
+	if (ept->cb.received) {
+		ept->cb.received(data, len, ept->priv);
+	}
+	return 0;
+}
+
+static struct rpmsg_mi_cb cb = {
+	.bound      = common_bound_cb,
+	.received   = common_recv_cb,
+};
+
+static int send(struct ipc_ept *ept, const void *data, size_t len)
+{
+	return rpmsg_mi_send(&ept->rpmsg_ep, data, len);
+}
+
+static int register_ept(struct ipc_ept **ept, const struct ipc_ept_cfg *cfg)
+{
+
+	if (!cfg || !ept) {
+		return -EINVAL;
+	}
+
+	int i;
+
+	for (i = 0; i < NUM_ENDPOINTS; ++i) {
+		if (!endpoints[i].name) {
+			struct rpsmg_mi_ctx_cfg ctx_cfg = {0};
+			struct rpmsg_mi_ept_cfg ept_cfg = {0};
+
+			endpoints[i].name       = cfg->name;
+			endpoints[i].ept.priv   = cfg->priv;
+			endpoints[i].ept.cb     = cfg->cb;
+
+			if (snprintf(endpoints[i].ipm_rx_name, IPM_CH_NAME_LEN, "IPM_%d",
+									MASTER ? 2*i : 2*i+1) < 0) {
+				goto cleanup;
+			}
+			if (snprintf(endpoints[i].ipm_tx_name, IPM_CH_NAME_LEN, "IPM_%d",
+									MASTER ? 2*i+1 : 2*i) < 0) {
+				goto cleanup;
+			}
```

Comment:
This code looks specific to IPM driver ipm_nrfx_icp. I think it should be generalized by taking IPM instance labels from devicetree or Kconfig.

Reply:
I think you’re right.